### ISSUE TITLE
Density of product of two uniform random variables evaluates to 0

### ISSUE THREAD

**Issue by ethankward** [sympy/sympy#17321]
```
>>> from sympy import *
>>> from sympy.stats import *
>>> x = Uniform('x', 0, 1)
>>> y = Uniform('y', 0, 1)
>>> density(x*y)
Lambda(_z, 0)
```

**Comment by ethankward** [comment]
A few months ago it was giving this result (commit a9853ab3d9b949be43ef577b54c1dc60823ad1cb):

```
>>> density(x*y)
Lambda(_z, Integral(Piecewise((0, (y >= 0) & (y <= 1) & (_z/y > 0) & (_z/y - 1 > 0)), (-Heaviside(0)/y, (y >= 0) & (y <= 1) & Eq(_z/y, 0) & (_z/y - 1 > 0)), (-1/y, (y >= 0) & (y <= 1) & (_z/y < 0) & (_z/y - 1 > 0)), (Heaviside(0)/y, (y >= 0) & (y <= 1) & (_z/y > 0) & Eq(_z/y - 1, 0)), (0, (y >= 0) & (y <= 1) & Eq(_z/y, 0) & Eq(_z/y - 1, 0)), ((Heaviside(0) - 1)/y, (y >= 0) & (y <= 1) & (_z/y < 0) & Eq(_z/y - 1, 0)), (1/y, (y >= 0) & (y <= 1) & (_z/y > 0) & (_z/y - 1 < 0)), ((1 - Heaviside(0))/y, (y >= 0) & (y <= 1) & Eq(_z/y, 0) & (_z/y - 1 < 0)), (0, True)), (y, 0, 1)))
```

**Comment by oscargus** [comment]
There is an issue with integrals as follows:
```
x = Symbol('x')
y = Symbol('y')
z = Symbol('z', real=True)
Integral(DiracDelta(z - y*x), (x, 0, 1)).doit()
Out[29]: 0

Integral(DiracDelta(z - x), (x, 0, 1)).doit()
Out[30]: -θ(-z) + θ(1 - z)
```
So if `x` is multiplied by a symbol the result of the integral is 0. It is possible to multiply by an integer and get something else. It doesn't matter if assumptions are put on `y`.

The result should probably be (θ = Heaviside):
```
   ⎛-z ⎞    ⎛    z⎞
  θ⎜───⎟   θ⎜1 - ─⎟
   ⎝ y ⎠    ⎝    y⎠
- ────── + ────────
    y         y    
```

which when integrated between 0 and 1 takes a very long time to finish...

**Comment by jksuom** [comment]
The denominator should probably be `|y|`.

**Comment by oscargus** [comment]
@jksuom correct. ( I used an integer and replaced it, hence, the neglectance...) I've noted that it does happen at times as well, not just here.

**Comment by oscargus** [comment]
It seems like setting either (non)positive or (non)negative conditions on `y` solves the problem. Unfortunately, it doesn't help to set it on the random variables as:
```
x = Symbol('x', nonnegative=True)
y = Symbol('y', nonnegative=True)
x = Uniform(x, 0, 1)
y = Uniform(y, 0, 1)
```
so not really sure where it does wrong. Still gives 0.

However,
```
x = Symbol('x', nonnegative=True)
y = Symbol('y', nonnegative=True)
z = Symbol('z', real=True)
integrate(DiracDelta(z- y*x), (x, 0, 1))
```
gives
![image](https://user-images.githubusercontent.com/8114497/63151605-2a190700-c00a-11e9-9524-bdb532249629.png)

(Note that the abs is gone since `y` is nonnegative, but for the general case it should be there.)

**Comment by jksuom** [comment]
The following lines will replace `DiracDelta` by a `Piecewise` that is zero (except at a single point).
https://github.com/sympy/sympy/blob/40da07b3a8ca3b315aaf25ec4ac21dfe4364a097/sympy/integrals/integrals.py#L495-L502
That happens when the integration variable `xab[0]` is real which is now always the case after commit a8e23375f. I think that the natural fix would be removing `DiracDelta._eval_rewrite_as_Piecewise` (which is useless anyway as `DiracDelta(0)` has no meaning).

**Comment by oscarbenjamin** [comment]
I hadn't seen this before:
```julia
In [8]: DiracDelta(x).rewrite(Piecewise)                                                                                          
Out[8]: 
⎧δ(0)  for x = 0
⎨               
⎩ 0    otherwise
```
That certainly looks like gibberish to me...

**Comment by oscargus** [comment]
`density(x*y)(z)` becomes
![image](https://user-images.githubusercontent.com/8114497/63529237-a75de380-c504-11e9-8f5b-ccd938a95aab.png)
when removing the rewrite.

**Comment by oscargus** [comment]
One reason it becomes a mess is that `Heaviside(0)` is undefined. With some manual steps I managed to get 
![image](https://user-images.githubusercontent.com/8114497/63532976-24408b80-c50c-11e9-81a0-7c7685458770.png)
 where probably the otherwise case should be zero (or rather, the first case should be for z >= 0.
![image](https://user-images.githubusercontent.com/8114497/63533143-83060500-c50c-11e9-9ff7-05e7894e3b7c.png)

Similarly for `density(x + y)(z)` I managed to get
`Piecewise((0, (y < z) & (z > y + 1)), (1, y < z), (-1, z > y + 1), (0, True))`
![image](https://user-images.githubusercontent.com/8114497/63533123-771a4300-c50c-11e9-82a4-f5e0a9d2de7f.png)

The problem is really the boundaries. Arbitrarily assigning `Heaviside(0)` to 0 or 1 when rewriting to Piecewise simplifies the problem tremendously.

**Comment by oscargus** [comment]
The density should probably be: `Piecewise((-log(Min(1, z)), z >= 0), (0, True))` with some manual simplifications.

**Comment by oscarbenjamin** [comment]
The OP example hangs on previous versions of sympy but this Integral result is a regression since 1.4:
```julia
In [1]: x = Symbol('x') 
   ...: y = Symbol('y') 
   ...: z = Symbol('z', real=True) 
   ...: Integral(DiracDelta(z - y*x), (x, 0, 1)).doit()                                                                                                       
Out[1]: 0
```

**Comment by asmeurer** [comment]
Bisected it to a8e23375f817258981ff29bb69cf62f2651ab4f5 CC @smichr

**Comment by oscarbenjamin** [comment]
That commit only exposes an underlying bug which was already present in 1.4. If x is declared as real then the integral evaluates to 0:
```
from sympy import *

x = Symbol('x', real=True)
y = Symbol('y')
z = Symbol('z', real=True)
print(Integral(DiracDelta(z - y*x), (x, 0, 1)).doit())
```
With that I get 0 on both 1.5 and 1.4.

The commit from @smichr there correctly replaces x with a symbol that has real assumptions if x is not declared as real.

So this is not really a regression since 1.4 but is a valid problem in integrals that predates 1.4. I will remove the milestone from this issue.
